ledger/acctupdates.go

```diff
@@ -896,28 +926,141 @@ func (au *accountUpdates) initializeCaches(lastBalancesRound, lastestBlockRound,
+		if flushIntervalExceed || loadCompleted {
+			// adjust the last flush time, so that we would not hold off the flushing due to "working too fast"
+			au.lastFlushTime = time.Now().Add(-balancesFlushInterval)
+
+			if !rollbackSynchronousMode {
+				// switch to rebuild synchronous mode to improve performance
+				au.dbs.Wdb.SetSynchronousMode(context.Background(), au.accountsRebuildSynchronousMode, au.accountsRebuildSynchronousMode >= db.SynchronousModeFull)
+
+				// flip the switch to rollback the synchronous mode once we're done.
+				rollbackSynchronousMode = true
+			}
+
+			// The unlocking/relocking here isn't very elegant, but it does get the work done :
+			// this method is called on either startup or when fast catchup is complete. In the former usecase, the
+			// locking here is not really needed since the system is only starting up, and there are no other
+			// consumers for the accounts update. On the latter usecase, the function would always have exactly 320 rounds,
+			// and therefore this wouldn't be an issue.
+			// However, to make sure we're not missing any other future codepath, unlocking here and re-locking later on is a pretty
+			// safe bet.
+			au.accountsMu.Unlock()
+
+			// flush the account data
+			au.committedUpTo(blk.Round())
+
+			// wait for the writing to complete.
+			au.waitAccountsWriting()
+
+			// The au.dbRound after writing should be ~320 behind the block round.
+			roundsBehind := blk.Round() - au.dbRound
+
+			au.accountsMu.Lock()
+
+			// are we too far behind ? ( taking into consideration the catchpoint writing, which can stall the writing for quite a bit )
+			if roundsBehind > initializeCachesRoundFlushInterval+basics.Round(au.catchpointInterval) {
+				// we're unable to persist changes. This is unexpected, but there is no point in keep trying batching additional changes since any futher changes
+				// would just accumulate in memory.
+				close(blockEvalFailed)
+				au.log.Errorf("initializeCaches was unable to fill up the account caches accounts round = %d, block round = %d. See above error for more details.", au.dbRound, blk.Round())
+				err = fmt.Errorf("initializeCaches failed to initialize the account data caches")
+				return
+			}
+
+			// and once we flushed it to disk, update the lastFlushedRound
+			lastFlushedRound = blk.Round()
+		}
+
+		// if enough time have passed since the last time we wrote a message to the log file then give the user an update about the progess.
+		if time.Now().Sub(lastProgressMessage) > accountsCacheLoadingMessageInterval {
+			// drop the initial message if we're got to this point since a message saying "still initializing" that comes after "is initializing" doesn't seems to be right.
+			select {
+			case skipAccountCacheMessage <- struct{}{}:
+				// if we got to this point, we should be able to close the writeAccountCacheMessageCompleted channel to have the "completed initializing" message written.
+				close(writeAccountCacheMessageCompleted)
```

Comment:
why do we close `writeAccountCacheMessageCompleted` here and in goroutine above? it will trigger "completed" message on exit but in fact we can fail just on another block applying iteration (since we are still in a loop at the moment of closing the channel).

Reply:
ahh. nice corner case.
I've added
```
if err == nil {
```
around the logging of the `completed` message, so that in case of an error, it wouldn't write that.